ouroboros-consensus/test-consensus/Test/Dynamic/General.hs

```diff
@@ -170,31 +173,70 @@ prop_general ::
      ( Condense blk
      , Eq blk
      , HasHeader blk
+     , RunNode blk
      )
   => SecurityParam
   -> TestConfig
   -> LeaderSchedule
   -> TestOutput blk
   -> Property
 prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} schedule
-  TestOutput{testOutputNodes} =
-    counterexample ("nodeJoinPlan: " <> condense nodeJoinPlan) $
-    counterexample ("schedule: " <> condense schedule) $
+  TestOutput{testOutputNodes, testOutputSlotBlockNos} =
     counterexample ("nodeChains: " <> unlines ("" : map (\x -> "  " <> condense x) (Map.toList nodeChains))) $
+    counterexample ("nodeJoinPlan: " <> condense nodeJoinPlan) $
     counterexample ("nodeTopology: " <> condense nodeTopology) $
+    counterexample ("slot-node-depth: " <> condense slotNodeDepths) $
+    counterexample ("schedule: " <> condense schedule) $
+    counterexample ("schedule': " <> condense schedule') $
+    counterexample ("consensus expected: " <> show isConsensusExcepected) $
     tabulate "consensus expected" [show isConsensusExcepected] $
     tabulate "shortestLength" [show (rangeK k (shortestLength nodeChains))] $
     tabulate "floor(4 * lastJoinSlot / numSlots)" [show lastJoinSlot] $
     tabulate "minimumDegreeNodeTopology" [show (minimumDegreeNodeTopology nodeTopology)] $
     prop_all_common_prefix
         maxForkLength
         (Map.elems nodeChains) .&&.
+    prop_all_growth .&&.
     conjoin
       [ fileHandleLeakCheck nid nodeInfo
       | (nid, nodeInfo) <- Map.toList nodeInfos ]
   where
     NumBlocks maxForkLength = determineForkLength k nodeJoinPlan schedule
 
+    -- remove entries from @schedule@ if any of:
+    --
+    -- * the node just joined in this slot (unless it's the earliest slot in
+    --   which any nodes joined)
+    --
+    -- * the node rejected its own new block (eg 'PBftExceededSignThreshold')
+    --
+    -- * the node forged an EBB
+    --
+    schedule' :: LeaderSchedule
+    schedule' =
+        Map.mapWithKey (filter . actuallyLead) `coerce` schedule
+      where
+        actuallyLead s cid = maybe False id $ do
+            let nid = fromCoreNodeId cid
+            let j   = nodeIdJoinSlot nodeJoinPlan nid
+
+            guard $ j < s || (j == s && isFirstJoinSlot s)
+
+            NodeOutput
+              { nodeOutputForges
+              , nodeOutputInvalids
+              } <- Map.lookup nid testOutputNodes
+            b <- Map.lookup s nodeOutputForges
+
+            pure $ not $
+              (||) (nodeIsEBB b) $
+              Set.member (blockPoint b) nodeOutputInvalids
+
+        isFirstJoinSlot s =
+            let NodeJoinPlan m = nodeJoinPlan
+            in
+            Just s == (snd <$> Map.lookupMin m)
```

Comment:
I'd just put this line on the same as `in`

Reply:
Changed to `where`.